lib/navigation-toc-postprocessor/extension.rb

```diff
@@ -0,0 +1,134 @@
+require 'asciidoctor/extensions' unless RUBY_ENGINE == 'opal'
+
+class NavigationTocPostprocessor < Asciidoctor::Extensions::Postprocessor
+
+  def process document, output
+    if document.attr 'navicons-toc'
+      html = NavIconsBackend.gen_nav_icons(document)
+      output = NavIconsBackend.insert_at(document, output, 'toc', html)
+    end
+    output
+  end
+end
+
+class NavIconsDocinfoProcessor < Asciidoctor::Extensions::DocinfoProcessor
+  # DocinfoProcessor as used in the chart-block-macro extension for adding HTML in head or footer
+  use_dsl
+  at_location :head
+
+  # inlusion of FontAwesome script, equal to the inclusion by the `:icons: font` attribute
+  FONTAWESOME_SCRIPT = '<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.5.0/css/font-awesome.min.css">'
+
+  # custom stylesheet for rendering the navigation icons
+  NAVICION_STYLESHEET = %(<style>
+/*navigation-toc-postprocessor asciidoctor extension*/
+div #toc-nav{
+font-size: 1.6em;
+display: flex;
+justify-content: space-around;
+padding-bottom: 10px;
+}
+</style>
+</head>)
+
+  def process doc
+    %(
+#{FONTAWESOME_SCRIPT}
+#{NAVICION_STYLESHEET})
+  end
+
+end
+
+class NavIconsBackend
+
+  def self.gen_icon(link="", fa_name='fa-home')
+    # filter link
+    # TODO check if this is consistent with asciidoctor.rb internal handling
+    url_parse = /link\:(.*?)\[(.*?)\]/.match(link)
+    ref_parse = /<<(([^#]*?)(\.\w+?)?(#(\w*?))?)(?:,(.*?))?>>/.match(link)
+    if url_parse
+      url = url_parse[1] || ""
+      desc = url_parse[2] || nil
+    elsif ref_parse
+      url = (ref_parse[2] + '.html') || ""
+      if ref_parse[4] # add attribute link
+        url << ref_parse[4]
+      end
+      desc = ref_parse[6] || nil
+    else
+      url = link
+      desc = nil
+    end
+
+    # render HTML
+    content = "<div id=\"toc-nav-item\">"
+    icon = "<i class=\"fa #{fa_name}\" aria-hidden=\"true\"></i>"
+    if url
+      content << "<a href=\"#{url}\""
+      if desc
+        content << " title=\"#{desc}\""
+      end
+      content << ">" + icon + "</a>"
+    else
+      content << icon
+    end
+    content << "</div>"
+
+    content
+  end
+
+  def self.gen_nav_icons(document)
+    content = ""
```

Comment:
In Asciidoctor html5 converter we are using an array to concat Strings:

```ruby
result = []

result << "<foo/>"
result << "<bar/>"

result * EOL
```

This should be better to read the source code of the generated HTML document otherwise everything will on one line.